pkg/parse/root.go

```diff
@@ -169,20 +169,35 @@ func (p *root) setSourceStatusWithRetries(ctx context.Context, newStatus sourceS
 		return status.APIServerError(err, "failed to get RootSync for parser")
 	}
 
+	currentRS := rs.DeepCopy()
+
 	setSourceStatus(&rs.Status.Source, p, newStatus, denominator)
 
-	continueSyncing := true
-	if rs.Status.Source.ErrorSummary.TotalCount > 0 {
-		continueSyncing = false
-	}
-	metrics.RecordPipelineError(ctx, configsync.RootSyncName, "source", rs.Status.Source.ErrorSummary.TotalCount)
+	continueSyncing := (rs.Status.Source.ErrorSummary.TotalCount == 0)
 	var errorSource []v1beta1.ErrorSource
-	if rs.Status.Source.Errors != nil {
+	if len(rs.Status.Source.Errors) > 0 {
 		errorSource = []v1beta1.ErrorSource{v1beta1.SourceError}
 	}
 	rootsync.SetSyncing(&rs, continueSyncing, "Source", "Source", newStatus.commit, errorSource, rs.Status.Source.ErrorSummary, newStatus.lastUpdate)
 
-	metrics.RecordReconcilerErrors(ctx, "source", status.ToCSE(newStatus.errs))
+	// Avoid unnecessary status updates.
+	if cmp.Equal(currentRS.Status, rs.Status, compare.IgnoreTimestampUpdates) {
+		klog.V(5).Infof("Skipping source status update for RootSync %s/%s", rs.Namespace, rs.Name)
+		return nil
+	}
+
+	csErrs := status.ToCSE(newStatus.errs)
+	metrics.RecordReconcilerErrors(ctx, "source", csErrs)
+	metrics.RecordPipelineError(ctx, configsync.RootSyncName, "source", len(csErrs))
```

Comment:
We still should record these two metrics even if there is no need to update the status.

Reply:
why? if they haven't changed the metric was recorded last time.

Reply:
Recording these metrics for every syncs helps understand questions like the successful rate of syncs in the last 30 days.

Reply:
These status update functions aren't being called after every sync tho. They're being called asynchronously every 5 seconds...

Reply:
You're right about rendering. `setRenderingStatus` is called in 3 different places after rendering (done, not done, and error). So it does get called after each render. But rendering happens every time the run loop triggers, which is on every poll, re-sync, or retry timer. So even if there's nothing to sync, the rendering still happens. Why would we want to send metrics for all of those if it just re-renders the same way it rendered last time?

`setSourceStatus` tho is only called when the source status changes, because it's being wrapped by `needToSetSourceStatus`. This extra check inside is to confirm that nothing has changed after converting the input status to the actual RSync API.

`SetSyncStatus` is also normally wrapped by `needToSetSyncStatus`, but that didn't get added to the async status updates. So the equality check in `SetSyncStatus` is covering that gap.

So even before I added the cmp.Equal checks it wasn't possible to use these metrics to compute "successful rate of syncs in the last 30 days". 

I'm also not sure that metric would actually be useful unless it was a ratio of commits that succeeded, not just a ration of attempts that succeeded. Otherwise the 1s retry timer would just bork the metric any time a broken commit was left broken.

Plus, if you use Prometheus without timestamps enabled, the metric would just appear unchanged every time you scrape it, with the latest value, unless it dropped off the 5m cache timeout. So you wouldn't be able to get that metric with Prometheus, only Cloud Monitoring, assuming it keeps track of the historical values every time they're reported. 

Plus metrics are designed to be lossy. If there's a network issue they don't usually retry. Otherwise you get cascading metric loss where a pile up of old metrics can stop new metrics from being reported. So you can't really produce a useful ratio metric over time unless the reporter is doing the math.

Reply:
I agree that today our code is not consistent on when to update status metrics for each component.
We should audit our metrics and make it consistent.